File: src/Tree/BalancedBinaryTree/BalancedBinaryTree.java
class Solution {
    final int UNBALANCED = -1;
    public boolean isBalanced(TreeNode root) {
        if(root == null) return true;
        return dfs(root) != UNBALANCED;
    }
    private int dfs(TreeNode root){
        //Base case: leaf node
        if(root.left == null && root.right == null) return 1;
        
        //Left Subtree DFS
        int leftH = 0;
        if(root.left != null){
            leftH = dfs(root.left);
        }
        if(leftH == UNBALANCED) return UNBALANCED;
        
        //Right Subtree DFS
        int rightH = 0;
        if(root.right != null){
           rightH = dfs(root.right);
        }
        if(rightH == UNBALANCED) return UNBALANCED;
        
        //Check current tree
        if(Math.abs(rightH - leftH) <= 1) return Math.max(rightH, leftH) + 1;
        return UNBALANCED;
    }
}

File: src/Tree/BalancedBinaryTree/BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    UNBALANCE = -1
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(root):
            if root is None:
                return 0
            
            left_height = dfs(root.left)
            
            right_height = dfs(root.right)
            
            if left_height == self.UNBALANCE or right_height == self.UNBALANCE:
                return self.UNBALANCE
            
            if 1 < abs(left_height - right_height):
                return self.UNBALANCE
            
            return max(left_height, right_height) + 1
        
        return dfs(root) != self.UNBALANCE
